frontend/catalyst/from_plxpr.py

```diff
@@ -351,7 +357,26 @@ def interpret_measurement_eqn(self, eqn: jax.core.JaxprEqn):
         )[0]
 
         primitive = measurement_map[eqn.primitive.name]
-        mval = primitive.bind(obs, shape=shaped_array.shape, shots=self._device.shots.total_shots)
+
+        # TODO: as we are in the process of migrating to dynamic measurement primitive shapes,
+        # we will gradually get rid of the shape argument for these primitives
+        # While we are in the migrating process, we need to handle them explicitly one by one
+        if primitive is sample_p:
+            mval = (
+                primitive.bind(obs, shots=self.device_shots, num_qubits=shaped_array.shape[1])
+                if isinstance(self.device_shots, int)
+                else primitive.bind(obs, self.device_shots, num_qubits=shaped_array.shape[1])
```

Comment:
When shots are concrete, we want to bind shots as kwarg so it is a integer paramater of the primitive instead of a `stablehlo.constant` SSA value. This is so that during lower, the return shape of `sample` can be directly lowered as `<concrete_shots x num_qubits x f64>`, instead of  `<? x num_qubits x f64>`.

The deeper reason for this distinction is that during `sample` bufferization, depending on whether the return type is static or dynamic, different allocations are necessary. For static shapes the allocation is `memref.alloc ()`, while for dynamic shapes the allocation is `memref.alloc (%size_to_be_allocated)`.

Reply:
> The deeper reason for this distinction is that during sample bufferization, depending on whether the return type is static or dynamic, different allocations are necessary. For static shapes the allocation is memref.alloc (), while for dynamic shapes the allocation is memref.alloc (%size_to_be_allocated).

Exactly, wouldn't that then get rid of the special handling for the two cases if it was always an SSA value?

Reply:
So we want the `sample` return shape to always be `<? x num_qubits x f64>`, and get rid of the static path altogether?

Reply:
Discussed, we decided to keep the static path